File: dromedary/src/decorator.rs
//! Shared forwarding helpers for Transport decorators, matching
//! dromedary/decorator.py.
//!
//! `forward_transport_all!($field)` emits a default forwarding impl for
//! every Transport method. Each individual forwarder is also exposed as
//! `fwd_<name>!` so decorators that want to override a few methods can
//! invoke only the forwarders they need instead of relying on skip
//! lists.

/// Build a decorator base URL by prefixing the inner transport's base.
/// `prefix` should include the trailing `+`, matching Python's
/// `_get_url_prefix()` convention (e.g. "fakenfs+").
pub fn prefixed_base(prefix: &str, inner: &dyn crate::Transport) -> ::url::Url {
    let inner_base = inner.base();
    let url = format!("{}{}", prefix, inner_base);
    ::url::Url::parse(&url).unwrap_or(inner_base)
}

#[macro_export]
macro_rules! fwd_external_url {
    ($field:ident) => {
        fn external_url(&self) -> $crate::Result<::url::Url> {
            self.$field.external_url()
        }
    };
}
#[macro_export]
macro_rules! fwd_can_roundtrip_unix_modebits {
    ($field:ident) => {
        fn can_roundtrip_unix_modebits(&self) -> bool {
            self.$field.can_roundtrip_unix_modebits()
        }
    };
}
#[macro_export]
macro_rules! fwd_is_readonly {
    ($field:ident) => {
        fn is_readonly(&self) -> bool {
            self.$field.is_readonly()
        }
    };
}
#[macro_export]
macro_rules! fwd_listable {
    ($field:ident) => {
        fn listable(&self) -> bool {
            self.$field.listable()
        }
    };
}
#[macro_export]
macro_rules! fwd_get {
    ($field:ident) => {
        fn get(
            &self,
            relpath: &$crate::UrlFragment,
        ) -> $crate::Result<Box<dyn $crate::ReadStream + Send + Sync>> {
            self.$field.get(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_has {
    ($field:ident) => {
        fn has(&self, relpath: &$crate::UrlFragment) -> $crate::Result<bool> {
            self.$field.has(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_stat {
    ($field:ident) => {
        fn stat(&self, relpath: &$crate::UrlFragment) -> $crate::Result<$crate::Stat> {
            self.$field.stat(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_clone {
    ($field:ident) => {
        fn clone(
            &self,
            offset: Option<&$crate::UrlFragment>,
        ) -> $crate::Result<Box<dyn $crate::Transport>> {
            self.$field.clone(offset)
        }
    };
}
#[macro_export]
macro_rules! fwd_abspath {
    ($field:ident) => {
        fn abspath(&self, relpath: &$crate::UrlFragment) -> $crate::Result<::url::Url> {
            self.$field.abspath(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_relpath {
    ($field:ident) => {
        fn relpath(&self, abspath: &::url::Url) -> $crate::Result<String> {
            self.$field.relpath(abspath)
        }
    };
}
#[macro_export]
macro_rules! fwd_put_file {
    ($field:ident) => {
        fn put_file(
            &self,
            relpath: &$crate::UrlFragment,
            f: &mut dyn ::std::io::Read,
            permissions: Option<::std::fs::Permissions>,
        ) -> $crate::Result<u64> {
            self.$field.put_file(relpath, f, permissions)
        }
    };
}
#[macro_export]
macro_rules! fwd_mkdir {
    ($field:ident) => {
        fn mkdir(
            &self,
            relpath: &$crate::UrlFragment,
            permissions: Option<::std::fs::Permissions>,
        ) -> $crate::Result<()> {
            self.$field.mkdir(relpath, permissions)
        }
    };
}
#[macro_export]
macro_rules! fwd_delete {
    ($field:ident) => {
        fn delete(&self, relpath: &$crate::UrlFragment) -> $crate::Result<()> {
            self.$field.delete(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_rmdir {
    ($field:ident) => {
        fn rmdir(&self, relpath: &$crate::UrlFragment) -> $crate::Result<()> {
            self.$field.rmdir(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_rename {
    ($field:ident) => {
        fn rename(
            &self,
            rel_from: &$crate::UrlFragment,
            rel_to: &$crate::UrlFragment,
        ) -> $crate::Result<()> {
            self.$field.rename(rel_from, rel_to)
        }
    };
}
#[macro_export]
macro_rules! fwd_set_segment_parameter {
    ($field:ident) => {
        fn set_segment_parameter(&mut self, key: &str, value: Option<&str>) -> $crate::Result<()> {
            self.$field.set_segment_parameter(key, value)
        }
    };
}
#[macro_export]
macro_rules! fwd_get_segment_parameters {
    ($field:ident) => {
        fn get_segment_parameters(
            &self,
        ) -> $crate::Result<::std::collections::HashMap<String, String>> {
            self.$field.get_segment_parameters()
        }
    };
}
#[macro_export]
macro_rules! fwd_append_file {
    ($field:ident) => {
        fn append_file(
            &self,
            relpath: &$crate::UrlFragment,
            f: &mut dyn ::std::io::Read,
            permissions: Option<::std::fs::Permissions>,
        ) -> $crate::Result<u64> {
            self.$field.append_file(relpath, f, permissions)
        }
    };
}
#[macro_export]
macro_rules! fwd_readlink {
    ($field:ident) => {
        fn readlink(&self, relpath: &$crate::UrlFragment) -> $crate::Result<String> {
            self.$field.readlink(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_hardlink {
    ($field:ident) => {
        fn hardlink(
            &self,
            rel_from: &$crate::UrlFragment,
            rel_to: &$crate::UrlFragment,
        ) -> $crate::Result<()> {
            self.$field.hardlink(rel_from, rel_to)
        }
    };
}
#[macro_export]
macro_rules! fwd_symlink {
    ($field:ident) => {
        fn symlink(
            &self,
            rel_from: &$crate::UrlFragment,
            rel_to: &$crate::UrlFragment,
        ) -> $crate::Result<()> {
            self.$field.symlink(rel_from, rel_to)
        }
    };
}
#[macro_export]
macro_rules! fwd_iter_files_recursive {
    ($field:ident) => {
        fn iter_files_recursive(&self) -> Box<dyn Iterator<Item = $crate::Result<String>>> {
            self.$field.iter_files_recursive()
        }
    };
}
#[macro_export]
macro_rules! fwd_open_write_stream {
    ($field:ident) => {
        fn open_write_stream(
            &self,
            relpath: &$crate::UrlFragment,
            permissions: Option<::std::fs::Permissions>,
        ) -> $crate::Result<Box<dyn $crate::WriteStream + Send + Sync>> {
            self.$field.open_write_stream(relpath, permissions)
        }
    };
}
#[macro_export]
macro_rules! fwd_delete_tree {
    ($field:ident) => {
        fn delete_tree(&self, relpath: &$crate::UrlFragment) -> $crate::Result<()> {
            self.$field.delete_tree(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_move {
    ($field:ident) => {
        fn r#move(
            &self,
            rel_from: &$crate::UrlFragment,
            rel_to: &$crate::UrlFragment,
        ) -> $crate::Result<()> {
            self.$field.r#move(rel_from, rel_to)
        }
    };
}
#[macro_export]
macro_rules! fwd_list_dir {
    ($field:ident) => {
        fn list_dir(
            &self,
            relpath: &$crate::UrlFragment,
        ) -> Box<dyn Iterator<Item = $crate::Result<String>>> {
            self.$field.list_dir(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_lock_read {
    ($field:ident) => {
        fn lock_read(
            &self,
            relpath: &$crate::UrlFragment,
        ) -> $crate::Result<Box<dyn $crate::lock::Lock + Send + Sync>> {
            self.$field.lock_read(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_lock_write {
    ($field:ident) => {
        fn lock_write(
            &self,
            relpath: &$crate::UrlFragment,
        ) -> $crate::Result<Box<dyn $crate::lock::Lock + Send + Sync>> {
            self.$field.lock_write(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_local_abspath {
    ($field:ident) => {
        fn local_abspath(
            &self,
            relpath: &$crate::UrlFragment,
        ) -> $crate::Result<::std::path::PathBuf> {
            self.$field.local_abspath(relpath)
        }
    };
}
#[macro_export]
macro_rules! fwd_get_smart_medium {
    ($field:ident) => {
        fn get_smart_medium(&self) -> $crate::Result<Box<dyn $crate::SmartMedium>> {
            self.$field.get_smart_medium()
        }
    };
}
#[macro_export]
macro_rules! fwd_copy {
    ($field:ident) => {
        fn copy(
            &self,
            rel_from: &$crate::UrlFragment,
            rel_to: &$crate::UrlFragment,
        ) -> $crate::Result<()> {
            self.$field.copy(rel_from, rel_to)
        }
    };
}

/// Forward every Transport method to `self.$field`. The caller must still
/// define `fn base(&self) -> Url`. Use this when no methods need to be
/// overridden; decorators that override a few methods should invoke the
/// individual `fwd_*!` macros instead.
#[macro_export]
macro_rules! fwd_all {
    ($field:ident) => {
        $crate::fwd_external_url!($field);
        $crate::fwd_can_roundtrip_unix_modebits!($field);
        $crate::fwd_is_readonly!($field);
        $crate::fwd_listable!($field);
        $crate::fwd_get!($field);
        $crate::fwd_has!($field);
        $crate::fwd_stat!($field);
        $crate::fwd_clone!($field);
        $crate::fwd_abspath!($field);
        $crate::fwd_relpath!($field);
        $crate::fwd_put_file!($field);
        $crate::fwd_mkdir!($field);
        $crate::fwd_delete!($field);
        $crate::fwd_rmdir!($field);
        $crate::fwd_rename!($field);
        $crate::fwd_set_segment_parameter!($field);
        $crate::fwd_get_segment_parameters!($field);
        $crate::fwd_append_file!($field);
        $crate::fwd_readlink!($field);
        $crate::fwd_hardlink!($field);
        $crate::fwd_symlink!($field);
        $crate::fwd_iter_files_recursive!($field);
        $crate::fwd_open_write_stream!($field);
        $crate::fwd_delete_tree!($field);
        $crate::fwd_move!($field);
        $crate::fwd_list_dir!($field);
        $crate::fwd_lock_read!($field);
        $crate::fwd_lock_write!($field);
        $crate::fwd_local_abspath!($field);
        $crate::fwd_get_smart_medium!($field);
        $crate::fwd_copy!($field);
    };
}

File: dromedary/decorator.py
# Copyright (C) 2006-2010 Canonical Ltd
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

"""Implementation of Transport that decorates another transport.

This does not change the transport behaviour at all, but provides all the
stub functions to allow other decorators to be written easily.
"""

from collections.abc import Iterator
from typing import IO, TYPE_CHECKING

from dromedary import (
    Transport,
    get_transport_from_path,
    get_transport_from_url,
    urlutils,
)

if TYPE_CHECKING:
    import os

    from dromedary import FileStream, Lock


class TransportDecorator(Transport):
    """A no-change decorator for Transports.

    Subclasses of this are new transports that are based on an
    underlying transport and can override or intercept some
    behavior.  For example ReadonlyTransportDecorator prevents
    all write attempts, and FakeNFSTransportDecorator simulates
    some NFS quirks.

    This decorator class is not directly usable as a decorator:
    you must use a subclass which has overridden the _get_url_prefix() class
    method to return the url prefix for the subclass.
    """

    def __init__(
        self,
        url: str,
        _decorated: Transport | None = None,
        _from_transport: "TransportDecorator | None" = None,
    ) -> None:
        """Set the 'base' path of the transport.

        :param _decorated: A private parameter for cloning.
        :param _from_transport: Is available for subclasses that
            need to share state across clones.
        """
        prefix = self._get_url_prefix()
        if not url.startswith(prefix):
            raise ValueError(
                f"url {url!r} doesn't start with decorator prefix {prefix!r}"
            )
        not_decorated_url = url[len(prefix) :]
        if _decorated is None:
            if urlutils.is_url(not_decorated_url):
                self._decorated = get_transport_from_url(not_decorated_url)
            else:
                self._decorated = get_transport_from_path(not_decorated_url)
        else:
            self._decorated = _decorated
        super().__init__(prefix + self._decorated.base)

    def abspath(self, relpath: str) -> str:
        """See Transport.abspath()."""
        return self._get_url_prefix() + self._decorated.abspath(relpath)

    def append_file(self, relpath: str, f: IO[bytes], mode: int | None = None) -> int:
        """See Transport.append_file()."""
        return self._decorated.append_file(relpath, f, mode=mode)

    def append_bytes(self, relpath: str, bytes: bytes, mode: int | None = None) -> int:
        """See Transport.append_bytes()."""
        return self._decorated.append_bytes(relpath, bytes, mode=mode)

    def _can_roundtrip_unix_modebits(self) -> bool:
        """See Transport._can_roundtrip_unix_modebits()."""
        return self._decorated._can_roundtrip_unix_modebits()

    def clone(self, offset: str | None = None) -> "TransportDecorator":
        """See Transport.clone()."""
        decorated_clone = self._decorated.clone(offset)
        return self.__class__(
            self._get_url_prefix() + decorated_clone.base, decorated_clone, self
        )

    def delete(self, relpath: str) -> None:
        """See Transport.delete()."""
        return self._decorated.delete(relpath)

    def delete_tree(self, relpath: str) -> None:
        """See Transport.delete_tree()."""
        return self._decorated.delete_tree(relpath)

    def external_url(self) -> str:
        """See dromedary.Transport.external_url."""
        # while decorators are in-process only, they
        # can be handed back into breezy safely, so
        # its just the base.
        return self.base

    @classmethod
    def _get_url_prefix(cls) -> str:
        """Return the URL prefix of this decorator."""
        raise NotImplementedError(cls._get_url_prefix)

    def get(self, relpath: str) -> IO[bytes]:
        """See Transport.get()."""
        return self._decorated.get(relpath)

    def has(self, relpath: str) -> bool:
        """See Transport.has()."""
        return self._decorated.has(relpath)

    def is_readonly(self) -> bool:
        """See Transport.is_readonly."""
        return self._decorated.is_readonly()

    def mkdir(self, relpath: str, mode: int | None = None) -> None:
        """See Transport.mkdir()."""
        return self._decorated.mkdir(relpath, mode)

    def open_write_stream(self, relpath: str, mode: int | None = None) -> "FileStream":
        """See Transport.open_write_stream."""
        return self._decorated.open_write_stream(relpath, mode=mode)

    def put_file(self, relpath: str, f: IO[bytes], mode: int | None = None) -> int:
        """See Transport.put_file()."""
        return self._decorated.put_file(relpath, f, mode)

    def put_bytes(self, relpath: str, bytes: bytes, mode: int | None = None) -> int:
        """See Transport.put_bytes()."""
        return self._decorated.put_bytes(relpath, bytes, mode)

    def listable(self) -> bool:
        """See Transport.listable."""
        return self._decorated.listable()

    def iter_files_recursive(self) -> Iterator[str]:
        """See Transport.iter_files_recursive()."""
        return self._decorated.iter_files_recursive()

    def list_dir(self, relpath: str) -> list[str]:
        """See Transport.list_dir()."""
        return self._decorated.list_dir(relpath)

    def _readv(
        self, relpath: str, offsets: list[tuple[int, int]]
    ) -> Iterator[tuple[int, bytes]]:
        """See Transport._readv."""
        return self._decorated._readv(relpath, offsets)

    def recommended_page_size(self) -> int:
        """See Transport.recommended_page_size()."""
        return self._decorated.recommended_page_size()

    def rename(self, rel_from: str, rel_to: str) -> None:
        """See Transport.rename."""
        return self._decorated.rename(rel_from, rel_to)

    def rmdir(self, relpath: str) -> None:
        """See Transport.rmdir."""
        return self._decorated.rmdir(relpath)

    def _get_segment_parameters(self) -> dict[str, str]:
        return self._decorated._segment_parameters

    def _set_segment_parameters(self, value: dict[str, str]) -> None:
        self._decorated._segment_parameters = value

    segment_parameters = property(
        _get_segment_parameters,
        _set_segment_parameters,
        doc="See Transport.segment_parameters",
    )

    def stat(self, relpath: str) -> "os.stat_result":
        """See Transport.stat()."""
        return self._decorated.stat(relpath)

    def lock_read(self, relpath: str) -> "Lock":
        """See Transport.lock_read."""
        return self._decorated.lock_read(relpath)

    def lock_write(self, relpath: str) -> "Lock":
        """See Transport.lock_write."""
        return self._decorated.lock_write(relpath)

    def _redirected_to(self, source: str, target: str) -> Transport:
        redirected = self._decorated._redirected_to(source, target)
        if redirected is not None:
            return self.__class__(self._get_url_prefix() + redirected.base, redirected)
        else:
            return None


def get_test_permutations() -> list[tuple[type, type]]:
    """Return the permutations to be used in testing.

    The Decorator class is not directly usable, and testing it would not have
    any benefit - its the concrete classes which need to be tested.
    """
    return []
